Add Quick-Access to favorite folder in left sidepanel in files-app

newhinton opened the pull request:
I have implemented the feature request from Issue #9714.
Fixes  #9714

The only thing which does not properly work, is the initial State of the checkbox which enables this functionality. It is properly set, but the initial state is always unchecked, regardless of the true value.

skjnldsv commented on apps/files/css/files.scss:
```diff
@@ -106,6 +106,19 @@
 	background-image: url('../img/delete.svg?v=1');
 }
 
+.nav-sidebar-spacer {
```
This should be standardise an put into the `apps.scss` file.
1. Please cut the code of the [settings-caption](https://github.com/nextcloud/server/blob/760b28ba4c7d4a9b6126aa4063238c2959eb263e/settings/css/settings.scss#L1244) class and put it into the `apps.scss` file (change the class name to something like `app-navigation-caption`
2. Update the uses of the old `settings-caption` across the server https://github.com/nextcloud/server/search?utf8=✓&q=settings-caption to you new class name
3. use the same class for your `Favorites` navigation element

Like so any other app that also register a spacer within the files app (from anywhere like their own application or from another nextcloud app) won't have anything to edit in the server core code :)

![capture d ecran_2018-06-03_18-56-38](https://user-images.githubusercontent.com/14975046/40889067-88e6d7f2-6760-11e8-8821-15a8cfd54bc6.png)

skjnldsv commented on apps/files/lib/Controller/ViewController.php:
```diff
@@ -169,18 +169,48 @@ public function index($dir = '', $view = '', $fileid = null, $fileNotFound = fal
+
+		$helper= new \OCA\Files\Activity\Helper($tagger);
+		$favElements = $helper->getFavoriteFilePaths($this->userSession->getUser()->getUID());
+		$favItems = $favElements["items"];
```
single quotes please ;)

skjnldsv commented on apps/files/lib/Controller/ViewController.php:
```diff
@@ -169,18 +169,48 @@ public function index($dir = '', $view = '', $fileid = null, $fileNotFound = fal
+		$favElements = $helper->getFavoriteFilePaths($this->userSession->getUser()->getUID());
+		$favItems = $favElements["items"];
+
+		$key="show_Quick_Access";
```
Same :p

skjnldsv commented on apps/files/lib/Controller/ViewController.php:
```diff
@@ -169,18 +169,48 @@ public function index($dir = '', $view = '', $fileid = null, $fileNotFound = fal
+		}
+
+		$i=0;
+		foreach($favElements["folders"] as $elem){
```
here also

skjnldsv commented on apps/files/lib/Controller/ViewController.php:
```diff
@@ -169,18 +169,48 @@ public function index($dir = '', $view = '', $fileid = null, $fileNotFound = fal
+		if($showFavoriteQuickAccess){
+			$nav->assign('favoritesItems', $favItems);
+			$nav->assign('favoritesFolders', $favFolder);
+			$nav->assign('setQuickAccessChecked', "checked");
```
Weee :see_no_evil:

skjnldsv commented on apps/files/templates/appnavigation.php:
```diff
@@ -1,21 +1,67 @@
-	<?php $pinned = 0 ?>
+		<?php
+		$pinned = 0;
+		$trashelement = null;
```
See other comment, we don't need to create specific entries for the trashbin, this should stay clean.

skjnldsv commented on apps/files/templates/appnavigation.php:
```diff
@@ -1,21 +1,67 @@
+		?>has-tooltip" title="<?php p($_['usage_relative'] . '%');
 		} ?>">
-			<a href="#" class="icon-quota svg">
+			<a href="#" class="nav-icon-quota svg">
```
I'm not sure the `nav-icon-quota` class [exists](https://github.com/nextcloud/server/search?q=nav-icon-quota&type=Code)? :thinking:

skjnldsv commented on apps/files/templates/appnavigation.php:
```diff
@@ -40,9 +86,14 @@ class="nav-icon-<?php p($item['icon'] !== '' ? $item['icon'] : $item['id']) ?> s
+			</div>
 			<label for="webdavurl"><?php p($l->t('WebDAV'));?></label>
-			<input id="webdavurl" type="text" readonly="readonly" value="<?php p($_['webdavurl']); ?>" />
+			<input id="webdavurl" type="text" readonly="readonly" value="<?php p(\OCP\Util::linkToRemote('webdav')); ?>" />
```
Why this change? Was it not working? (I did not checked) :)

skjnldsv reviewed (changes requested):
Hello! Thanks for your pull request and welcome! :) :tada: 

I added quite a few comments to explain some of the required changes :)
Feel free to ask for more details if I did not explained properly ;)

newhinton commented:
@skjnldsv The quotes and the webdav/quota-changes are fixed, they were made by mistake, thanks for pointing out! What do you mean with "put the css in app.scss"? Aren't they? Or do you mean that the name of my style is misleading and should be changed to something more descriptive?

The other changes will need some time since they are not quick fixes.

skjnldsv commented:
> What do you mean with "put the css in app.scss"? Aren't they?
> Or do you mean that the name of my style is misleading and should be changed to something more descriptive?

Sorry, I should have been more clear :)
The /core/css/apps.scss file contains the design standards across all of our apps and core. That includes (but not limited to) the app navigation, the app settings, the popoverenus, the app-content-lists... etc. Since we already have a design for the spacer (see the settings section of nextcloud and my screenshot) we should use this design but make it available for everyone in the future. So move the `settings-caption` property into the app-navigation part of the apps.scss file and rename it to a class that fit the class name pattern we use (lile app-navigation-caption for example :) )
https://github.com/nextcloud/server/blob/760b28ba4c7d4a9b6126aa4063238c2959eb263e/settings/css/settings.scss#L1244-L1257

Therefore, since your pull request will most likely be integrated into 14, every app that want to add a caption-like design into their navigation will be able to use our design thanks to you :wink:

skjnldsv commented:
> The other changes will need some time since they are not quick fixes.

It shouldn't be that much, don't worry. Since we already have the code to add items to the app navigation, just make your script add the favorites items with the `\OCA\Files\App::getNavigationManager()->add()` [function](https://github.com/nextcloud/server/blob/6ea444b24387e9b516a44000df0db0f6af786f95/apps/files/lib/Controller/ViewController.php#L162) and the spacer with the same way.

You may need to edit the `add` [function](https://github.com/nextcloud/server/blob/cabd51442543d4cc2aa71b9f9d2291220d300553/lib/private/NavigationManager.php#L87) to add a new property like `type` to differentiate the 'traditional' navigation sitems from the spacers ;)

skjnldsv commented on core/css/apps.scss:
```diff
@@ -561,6 +561,25 @@ kbd {
 		list-style-type: none;
 	}
+
+	.settings-caption {
```
`app-navigation-caption` would be better please :)

newhinton replied:
Yeah, you are right about that, forgot to change its name after getting it to work :D


But there is still a last todo on my list, the Checkbutton wich enables this functionality for the user is not properly set on page load. It works and sets the correct value in the backend, but the current state is not applied on page load. Where do i need to set this value initially?

newhinton commented:
Yeah, you are right about that, forgot to change its name after getting it to work :D

But there is still a last todo on my list, the Checkbutton wich enables this functionality for the user is not properly set on page load. It works and sets the correct value in the backend, but the current state is not applied on page load. Where do i need to set this value initially?

skjnldsv commented on apps/files/templates/appnavigation.php:
```diff
@@ -40,9 +42,14 @@ class="nav-icon-<?php p($item['icon'] !== '' ? $item['icon'] : $item['id']) ?> s
 				<label for="showhiddenfilesToggle"><?php p($l->t('Show hidden files')); ?></label>
 			</div>
+			<div id="files-setting-showFavoriteQuickAccess">
+				<input class="checkbox" id="showfavoritequickaccessToggle" checked="checked" type="checkbox">
```
> But there is still a last todo on my list, the Checkbutton wich enables this functionality for the user is not properly set on page load. It works and sets the correct value in the backend, but the current state is not applied on page load. Where do i need to set this value initially?

The status is always checked by default here.
Either you set the state on javascript on load (not quite fond of this), or you pass it via php using the `getUserValue` config check (Better solution I think)

skjnldsv replied:
This is looking great by the way! :wink:

newhinton replied:
Yes, but the checked="checked" does not change the state. It is always disabled by default, and i think it is set by javascript to unchecked. I have tried to mimic the behaviour of the showHiddenFiles-Toggle, but it does not seem to work. I don't now enough about the internal workings of the core, so i'm stuck here

And thank you for the compliment  ;)

skjnldsv replied:
You can check the source code to see if php is properly setting this up.
But like Jan said, don't bother with the settings, remove it, it will be easier :)

And it's a great addition, so I guess less work from you and more pleasure for the users ;)

jancborchardt commented:
Good stuff! See my design comments in the issue at https://github.com/nextcloud/server/issues/9714 :)

skjnldsv commented on core/css/apps.scss:
```diff
@@ -561,6 +561,41 @@ kbd {
+		pointer-events:none;
+	}
+
+	.app-navigation-subelement {
```
Okay, I get it now! :)
There is already a standard for this:
Look at our documentation : https://docs.nextcloud.com/server/13/developer_manual/design/navigation.html

skjnldsv reviewed (changes requested):
Amazing work! Let's review this :)
- The Favourite first item menu is not collapsible, this should be possible
- The three-dot-more-icon should be hidden until the user have at least one directory favourited 
- The list is not updated in real time

That's all I found so far :)
Thanks a lot!

newhinton commented:
What do you mean with "The Favourite first item menu"?

The list cannot be updated in a responsive way, since it's nearly fully created in the server and then served to the user. Making it possible to refresh it in realtime would require to completely rewrite the navbarcode, to a responsive way. That is not something i intent to do in the near future, nor do i beliefe that it belongs in this pull request.

skjnldsv commented:
> What do you mean with "The Favourite first item menu"?

I mean the top level menu. The entry that says "Favorites" :)

skjnldsv commented:
@newhinton Sure :)
https://github.com/nextcloud/server/blob/f347e2e4a6cab71c35d58ff21692e23d7443db60/apps/files/js/tagsplugin.js#L89

skjnldsv commented on apps/files/js/navigation.js:
```diff
@@ -146,11 +146,17 @@
-				var dotmenu = document.getElementById("quickaccess-list");
-				dotmenu.style.display='';
+
+				if(!$("#favorites-toggle" ).hasClass('open')){
```
`toggleClass` 
https://api.jquery.com/toggleClass/

newhinton commented:
@skjnldsv check it out now, it should be pretty finished now

skjnldsv reviewed (approved):
Beautiful! :)

MorrisJobke commented on apps/files/js/navigation.js:
```diff
@@ -127,9 +136,205 @@
+					scope.reverse(list);
+				}
+			});
+			*/
```
What is this code about? Could it be removed or fixed?

newhinton replied:
This is now removed

MorrisJobke commented on apps/files/lib/Controller/ViewController.php:
```diff
@@ -159,28 +159,100 @@ public function index($dir = '', $view = '', $fileid = null, $fileNotFound = fal
+
+
+		//Get Favorite-Folder
+		$tagger=\OC::$server->getTagManager();
```
Please inject it via the constructor. Our DI container then takes care of providing it.

MorrisJobke commented on apps/files/lib/Controller/ViewController.php:
```diff
@@ -159,28 +159,100 @@ public function index($dir = '', $view = '', $fileid = null, $fileNotFound = fal
+		$user = $this->userSession->getUser()->getUID();
+
+		//Load QuickAccess-Defaults
+		$sorting=$this->config->getUserValue($user,$this->appName,'quickaccess_sorting_strategy','date');
```
Could you also fix the code style? We us spaces around `=` and a space after each `,`.

MorrisJobke commented on apps/files/lib/Controller/ViewController.php:
```diff
@@ -159,28 +159,100 @@ public function index($dir = '', $view = '', $fileid = null, $fileNotFound = fal
+		//Load QuickAccess-Defaults
+		$sorting=$this->config->getUserValue($user,$this->appName,'quickaccess_sorting_strategy','date');
+		$reverseListSetting=$this->config->getUserValue($user,$this->appName,'quickaccess_reverse_list','false');
+		if($this->config->getUserValue($user,$this->appName,'show_Quick_Access',true)){
```
Also there are spaces like this `if (` and `) {` - same goes for the else: `} else {`

MorrisJobke commented on apps/files/templates/appnavigation.php:
```diff
@@ -1,19 +1,71 @@
+			<li data-id="<?php p(isset($item['href']) ? $item['href'] : $item['id']) ?>"
+				class="nav-<?php p($item['id']) ?> <?php p($item['classes']) ?> <?php p($pinned===1?'first-pinned':'') ?> <?php if($item['enableQuickaccess']=='true'){ ?> open<?php } ?>"
+				<?php if(isset($item['folderPosition'])){ ?> folderPos="<?php p($item['folderPosition']);?>"<?php } ?>
+				<?php if($item['id']=='favorites'){?>id="favorites-toggle"<?php } ?>>
```
Could we use the identity check here? `===` instead of `==` and also spaces around the operator.

MorrisJobke reviewed (dismissed):
I added some code style nitpicks. Could you check them and fix them? Some rough guidelines can be found at https://docs.nextcloud.com/server/13/developer_manual/general/codingguidelines.html#php

juliushaertl commented on apps/files/appinfo/routes.php:
```diff
@@ -76,6 +76,36 @@
 				'url' => '/ajax/getstoragestats.php',
 				'verb' => 'GET',
 			],
+			[
```
I'd prefer to prefix the routes, so it is clear that they are related to the favorites:
```
/api/v1/quickaccess/show
/api/v1/quickaccess/hide
/api/v1/quickaccess/setSorting
```

newhinton replied:
I have changed the endpoints

juliushaertl commented:
Some comments about the settings there:
![image](https://user-images.githubusercontent.com/3404133/41543866-0b265ac2-7318-11e8-969c-447d2e922f58.png)

- The first checkbox can be removed, as the state is already triggered by clicking the toggle indicator in the sidebar
- The sorting entries should be options not checkboxes, since otherwise users will expect that they can all be checked. But there is only an option for one at a time.
- I'm still not sure if the sorting options aren't to much there. @nextcloud/designers

newhinton commented:
@juliushaertl I agree about your first and second argument, ill change it accordingly. But i disagree with your third argument. The list is quiet unstructured initially, because it is sorted by date added. If the user wishes to sort it by alphabet, he can easily do so. But making this default, and not toggleable could be bad, because displaying it by date added could be useful for users because it allows them a customized list, until this is implemented in the future as a standalone feature

newhinton commented:
After some testing, i have found a minor bug. If htacces.RewriteBase is set, my linkgeneration does not work (correctly).
They still link to the correct page, and nothing changes, at least not in raw functionality.
Only the url switches from 
http://192.168.225.4/cloud_/_apps/files/?dir=/&view=favorites
to
http://192.168.225.4/cloud_/index.php_/apps/files/?dir=/Family%20Photos&fileid=9

I would expect the url to be automatically rewritten, and in my main instance i can observe this intended behaviour. I am not sure if this is a bug, or if something in my development-instance is not fully set up. 

Long story short; 

is there a better way to get the weblink than this approach:

$link=\OC::$WEBROOT.'/index.php/apps/files/?dir='.$elem;

 https://github.com/newhinton/server/blob/084f1a551672c113597dd2670bb43110e94aba6c/apps/files/lib/Controller/ViewController.php#L216-L218

skjnldsv commented:
> is there a better way to get the weblink than this approach:
> $link=\OC::$WEBROOT.'/index.php/apps/files/?dir='.$elem;

Use the urlGenerator. https://docs.nextcloud.com/server/13/developer_manual/api/OC/URLGenerator.html?highlight=urlgenerator#OC\URLGenerator::linkToRouteAbsolute

`$this->urlGenerator->linkToRouteAbsolute('files.view.index', ['dir' => $elem])`

newhinton commented:
Per [Nextcloud - Doc](https://docs.nextcloud.com/server/13/developer_manual/general/codingguidelines.html#license-headers) i have added a copyright header, if it was the wrong place or if it is unnessessary please say so.

skjnldsv commented:
> Per Nextcloud - Doc i have added a copyright header, if it was the wrong place or if it is unnessessary please say so.

Nope, it's perfect :)
You could add your name to every file you edited as well.

By the way, I added the `app-navigation-caption` in another pr that got merged as well.
Please drop your implementation as I fixed some stuff already and rebase! ;)
https://github.com/nextcloud/server/commit/e01f004a130cc305ba7dcd8f7b82a8867ee54462

newhinton commented:
The sorting-option "date" is currently broken until i find out why the array which returns the folderlist is now sorted by alphabet.

jancborchardt commented:
@newhinton @juliushaertl I’d say all those settings should actually not be in there. Sorting settings for the favorites are a bit overwhelming. Let’s just sort them by last modified up top, so then you have the most relevant stuff visible.

Also, the word "Quickaccess" should not be used _anywhere_ in the interface. It’s basically favorites, so that’s what it should be called, if anything. ;)

newhinton commented:
@jancborchardt The quickaccess was already removed as @juliushaertl mentioned earlier that it was indeed a duplicate.

I have created a toggle to hide away the dot-menu, so that the user is not confronted with it if he does not want to be. (It is disabled by default). This should allow more advanced users to use that functionality while keeping the ui as clean as possible

jancborchardt commented:
@newhinton ok – but where would that toggle be then, in the config.php? Remember that we want to only introduce settings in the interface where absolutely necessary – and this seems not like it.

Even a config.php setting needs to be maintained, so I would strongly advise just sticking to one configuration, which in this case would be sorting by recently modified up top.

newhinton commented:
@jancborchardt The toggle (a checkbox) is located under the appsettings at the bottom of the navbar. Directly below the checkbox for hidden files. This should move the Toggle away from "the ui" (It's still in the ui, but not in the part that is always or often presented to the user) and keep it clean, while still maintaining the ability to toggle it. I think this solution is a good balance between keeping the functionality usable and keeping as much out of the way as possible.

jancborchardt commented:
@newhinton sorry, but this option is not nearly of the same importance as the global setting to "Show hidden files". I need to insist that we remove it from there. :wink:

Otherwise we sooner or later end up with [this](https://blog.codinghorror.com/this-is-what-happens-when-you-let-developers-create-ui/):
![](https://blog.codinghorror.com/content/images/uploads/2006/11/6a0120a85dcdae970b0128776fec64970c-pi.png)

And that is why we have the [design guideline](https://nextcloud.com/design/) to »Do things automatically instead of offering configuration options.« – a great related read is [»Choosing our Preferences«](https://ometer.com/preferences.html).

newhinton commented:
@jancborchardt I have a solution for this problem. We use a value in the config.php, with an non binary value.
This allows us to do multiple things: 
Turn off quickaccess fully (State 0), if not wanted by the admin, 
enable it (which should be the default (State 1)) 
and enable additional sorting (State 2)
It could also be expanded later, if a custom ordering of the favorites is implemented.


This removes the additional checkbox from the ui, but keeps the functionality, where the admin can configure how much freedom the user has.

MorrisJobke commented:
> This removes the additional checkbox from the ui, but keeps the functionality, where the admin can configure how much freedom the user has.

The feature should either be there or not. We had the same with so many other things in the past. Avatars were optional or file locking was optional for example. Both of them resulted in glitches caused by wrong defaults or that just a handful of people (literally) used. THat's the reason we want to go for sane defaults for everyone. If it turns out to have really an urge for a config switch we can think about it. But for now I would also vote for a feature that just does work for everybody in the same way.

jancborchardt commented:
> We use a value in the config.php, with an non binary value.

Then we end up with 3 more config states to always test and maintain. You see the problem. ;)

(The config file is not some kind of kitchen sink for all kinds of settings – even that we should keep clean.)

newhinton commented:
@jancborchardt @skjnldsv Originally i wanted to implement custom ordering, where the user can rearrange the folders as he wished (like most fileexplorers allow). Initially i thought this was to difficult to begin with, but after some research i found out that it is quiet easy. (Until i implemented this, sorting should take its place)

Is this something i should do? It does not require any toggles or boxes or stuff, which is what we want

skjnldsv commented:
Okay, the appnavigation.php definitely needs to be standardized.
the idea is to have a pattern and not some ifs conditions :)

I'm suggesting adding a support for an array first.

Right now we only have one level allowed into `$_['navigationItems']`. Add a condition that checks if one of the item have the key `$item['children']` and if so, loop over the same function again on those items.
Add the proper checks and you're good to go!

Another step is to add the support of the menu.
Idem, don't add conditions on the appNavigation template, just use a new key like `$item['utils']['actions']` and loop over this array to add the proper popovermenu and items.

You can see a similar implementation of the menu in javascript here: https://github.com/nextcloud/server/blob/master/settings/src/components/appNavigation.vue
https://github.com/nextcloud/server/blob/master/settings/src/components/appNavigation/navigationItem.vue

newhinton commented:
@skjnldsv You were faster ;) 

>The current way sublist-handling works is the following:
>
>For every item in $_['navigationItems']
>add an <li></li> element with the proper attributes. 
>if current element of $_['navigationItems'] has the favorites-id, then add a single <ul>-tag
>
>continue to add listelements (which are then added as sublists)
>if current element of $_['navigationItems'] has the "end"-value, then add the closing </ul>-tag
>(which then ends the sublist)


This was my initial idea (which keeps the existing structure intact), but a different way to tackle this would be your arrays.

But i am not sure what you mean with 'function', should i create literally a function which then adds a single <li></li> element to the structure and only call that function in a loop (and its recursive sublist-counterparts)?

skjnldsv commented:
Yes, basically this needs to be put into a function so you can call the function within itself
https://github.com/newhinton/server/blob/81e651d0f88ba43f39fecb5b4993abd1d331cc97/apps/files/templates/appnavigation.php#L8-L13

newhinton commented:
Okay, then i understood what you meant :) I'll tackle it

jancborchardt commented:
> Originally i wanted to implement custom ordering, where the user can rearrange the folders as he wished

Let’s first go with the simplest way. :) Also – custom ordering brings a bunch of issues, like accidental drags (especially when wanting to scroll on mobile), and confusion about what the sorting is then.

newhinton commented:
@jancborchardt It seems that HTML5 does not allow drag&drop on li-tags by default, you would need to actively implement it in jquery, so accidents are not an issue.
And i don't think that it's not that confusing to have a custom ordering, as soon as the user drags them around, he knows that it is custom, and if he never does that, it's fine too since it is then sorted by last favorite added.

jancborchardt commented:
@newhinton in any case – let’s first simply start with sorting by last modified first. Everything step by step. :)

newhinton commented:
@skjnldsv Since you are the files-pro here, how do i get the last modified timestamp to a given foldername? Is there an easy way?

newhinton commented:
@skjnldsv Where should i document which values are need to be set in order to get a sublist and how to build one?

skjnldsv commented:
@newhinton i have no idea for the timestamp, it depends on what data are served I guess. Sorry, files is not my specialty though! :(

Nextcloud have a repo for the documentation if you want. :)

newhinton commented:
@jancborchardt @skjnldsv I have implemented multiple sorting strategies. The currently active is custom ordering, with working drag&drop. Collapsed-state is stored, aswell as the custom order. The appnavigation.php uses recursive functions now, which allows easy adding of sublistelements to any navbar entry. 

At the moment sorting by last date modified  is only the fallback, which means you can test everything with only changing one line of code:

ViewController.php:219 : draggableSublist needs to be 'false' to disable d&d

If you want do change the fallback, you need to change either

navigation.js:52 or navigation.js:184 to the proper ordering.

skjnldsv commented on apps/files/templates/appnavigation.php:
```diff
@@ -23,26 +22,102 @@ class="nav-icon-<?php p($item['icon'] !== '' ? $item['icon'] : $item['id']) ?> s
+ *
+ * @return int Returns the pinned value
+ */
+function NavigationListElements($item, $pinned, $l) {
```
$pinned default false?
Also please use strict types :)
Should probably go as last parameter?

newhinton replied:
What do you mean with $pinned default false?  You mean the 67. Line? This is the default line from the current version in the server, apart from moving it to the function, i did not change it.

I have not used strict types in php before, how would i do that?

I agree with you setting pinned as the last parameter

skjnldsv commented:
Tests are not happy :)

jancborchardt commented:
@newhinton the default sorting should be by last modified and not the custom sorting, as said. ;)

@skjnldsv what do you think about the flash before it sorts? Can't we sort it server-side?

newhinton commented:
@jancborchardt Why do you think that lastmodified is better than a custom order? I think the ability to customize the list results in a better feature, since the user can customize it to his own workflow. It is also what a user would expect, since most explorer handle bookmarks the same way, and while it is slightly more complex to maintain than a fixed sorting, the cost far outweigh the benefits. 

Also the user than has to deal with an everchanging ui-element, which changes more or less randomly, because a file could change without the user noticing (because a file was uploaded by a third user, or an automated sync). I regard this behavior as highly confusing and not really transparent to the user. 

We could introduce the ability to change sorting, but then we need to discuss the "where to store settings" part again, which i dont want because we already have a better, more transparent solution at hand. (And we agreed to not introduce more settings because of ui-clutter and etc.)

I cant really see where the user would benefit from a changing ui where he needs to find what he wants everytime again because it moved.

newhinton commented:
Todo: Fix bug where on mobile the sidebar closes when the collapsed-button is triggered

skjnldsv commented:
@newhinton where are we here?

newhinton commented:
@skjnldsv
I have found a bug where on mobile the navbar would close completely when the collapsedstate is changed

Funtional state:
The feature is complete and contains all the nessessary parts

Technical state:
the beforementioned bug

Style-state:
I should check everything again, just in case something slipped my view

and the whole how to sort discussion is still going

newhinton commented:
@skjnldsv I have fixed the bug and therefore i thing i have finished the feature. 

But can you give me feedback on the animation i added while dragging an item? It gets a different icon when dragged arount to make the items visible which can be sorted.

jancborchardt commented:
@newhinton

> Can you explain why you think that lastmodified is better than a custom order?

- Because we need a default anyway
- Because it's less code to maintain
- Because it's a good first step
- Because it will always put the relevant stuff up top

So please – while we can do further things in future pull requests – let's keep the first one simple. We work in small steps and iterations to prevent overengineering or needing to maintain stuff no one uses. :)

newhinton commented:
The default would be dateadded, which would be the good first step you mentioned. We have already passed that after introducing sorting in any way.

The difference in code to maintain is not really that big, counted ~65 Lines of code more in the custom order, which includes brackets and so on.

Last Date modified will NOT always show relevant stuff, since a cloud is often a shared medium (shared by bots which sync stuff, friends moving files in your shared folders, etc.). This will often result in an order which is not understandable by the user. It is unclear why it changed, since there is no explanation how it sorts.

We dont need to make it discoverable, if the user never detects it, he can just use last date added, which does not have the ability to change randomly.

A user could drag a folder into another folder, which could have WAY more serious consequences (synced folders, dataleak in shared folders, etc). I do not consider accidential dragging of elements which pose no risk at all (except from a messed up sorting, which can be solved by dragging it back)

And as i stated earlier, dragging on mobile is not possible. You simply cannot drag on mobile browsers, if you do not implement it specifically.

You dont need a default, because it's a custom order. 



From a feature-based standpoint, i consider this feature ready. I have implemented what i intended to achieve here, which will add benefit to users if they decide to use it. I do not consider "lastdatemodified" as a good design choice, since it introduces intransparency and confusion to a less savy user. Also it limits the benefits since you would always need to find your folders again. 

As long as there are no technical problems, bugs or nessessary codestylechanges, i consider the work finished, and i will not change the sorting since then i would have developed a feature which i would never use.
I have explained how you can change the sorting-order, if you really wish to change it, please do it yourself.

skjnldsv reviewed (changes requested):
- Adding a file to favorite create a new entry in the left menu with the folder icon despite the fact that the starred file is not a directory
- On the favorite menu, when opening a folder details sidebar, clicking the star keeps adding entries to the menu and do not remove the directory from the favourites

newhinton commented:
@jancborchardt What about a dropdown menu? I know, its still a setting more in the ui, its still a little bit more to maintain, but i'm trying to make it work for everyone. Also it is a more slender solution to the sorting-problem as the last time. Your thoughts about that?

newhinton commented:
@skjnldsv I should have fixed the bugs

skjnldsv commented on apps/files/templates/appnavigation.php:
```diff
@@ -17,7 +17,7 @@ class="pinned <?php p($pinned === 0 ? 'first-pinned ' : '') ?><?php
 				<p id="quotatext"><?php
 					if ($_['quota'] !== \OCP\Files\FileInfo::SPACE_UNLIMITED) {
-						p($l->t('%s of %s used', [$_['usage'], $_['total_space']]));
+						p($l->t('%s of %s', [$_['usage'], $_['total_space']]));
```
Ah no! Do it in another pr please :)
Otherwise it's confusing!

newhinton replied:
Small missunderstanding ;)

skjnldsv reviewed (approved):
Ok from me.
Please rebase to master.

Only downside is that clicking a favourited folder change the whole page instead of loading the folder only. But we can fix this in a later pr.

skjnldsv commented:
Sigh... new pr is here: #10176 :disappointed:

DarkknightAK commented:
How can I pin an folder to the left sidebar?

newhinton commented:
@DarkknightAK
What do you mean?
if you want to use the quickaccess list, you can simply mark a folder as a favorite, that adds the folder there

DarkknightAK commented:
I'm thinking more on something like you know from any file manager where you can pin folders to the sidebar.

![NC_sidebar](https://user-images.githubusercontent.com/8517736/71577501-af521380-2af4-11ea-89ea-78686e49c9da.png)

and with folders I also mean labels. so you can show in addition to the predefined stuff in the sidebar stuff that was pinned by the user.

DarkknightAK commented:
Cause now you have 5 predefined items in the sidebar and you need two clicks (favorites -> folder) you come to your favorite folders. For tags you need three clicks (Tags drop down tag search and than have the results to click)

newhinton commented:
Ah okay

that is not possible at the moment. The only way to add a folder to the sidebar is to mark it as a favorite

however, you can expand the favorites in the sidebar, this will allow you to get to a folder quickly. it should stay expanded over sessions

DarkknightAK commented:
Ah cool that's fine to me. Thanks.

would it be also possible to pin labels that way to have an expand at labels and have there the labels.

newhinton commented:
Probably yes, but it needs to be developed and accepted. Also since this pr is already closed, could you open a new issue for that?

DarkknightAK commented:
#18607
